Added “SemanticModel.WithSuppressAccessChecks” extension method.

OmerRaviv opened the pull request:
The `WithSupressAccessChecks` method creates a SemanticModel that ignores accessibility rules when answering semantic questions.

This allows consumers to ask semantic questions using the same semantic rules as the ones used in debugger’s Expression Evaluator, where accessibility rules do not matter. This API is an absolute necessity for 3rd parties (such as OzCode) who want to create debugger-related productivity tools on top of Roslyn.

Added unit tests for C# and VB.NET to cover both regular and speculative analysis.

fixes #378

<!---
@huboard:{"order":514.0,"milestone_order":26,"custom_state":""}
-->

gafter commented on src/Compilers/Core/Portable/Compilation/SemanticModel.cs:
```diff
@@ -63,6 +63,16 @@ public SyntaxTree SyntaxTree
         /// <summary>
+        /// Returns true if this is a SemanticModel that ignores accessibility rules when answering semantic questions.
+        /// </summary>
+        public bool IsSuppressingAccessChecks { get; protected set; }
```
The setter should be internal.

gafter commented on src/Compilers/VisualBasic/Portable/Compilation/SemanticModel.vb:
```diff
@@ -106,6 +106,14 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
         Friend MustOverride Function GetCrefReferenceSymbolInfo(crefReference As CrefReferenceSyntax, options As SymbolInfoOptions, Optional cancellationToken As CancellationToken = Nothing) As SymbolInfo
 
+        Public Overrides Function WithSuppressAccessChecks() As SemanticModel
+            Dim moduleSymbol = TryCast(Compilation.SourceModule, SourceModuleSymbol)
```
If the current model is already ignoring accessibility, we should just return it.

gafter commented on src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:
```diff
@@ -51,6 +51,11 @@ internal abstract class CSharpSemanticModel : SemanticModel
+
+        public override SemanticModel WithSuppressAccessChecks()
+        {
+            return new SyntaxTreeSemanticModel(Compilation, this.SyntaxTree, suppressAccessChecks: true);
```
If the current model is already suppressing the checks, we should just return it.

AnthonyDGreen commented on src/Compilers/Core/Portable/Compilation/SemanticModel.cs:
```diff
@@ -63,6 +63,16 @@ public SyntaxTree SyntaxTree
+        /// <summary>
+        /// Creates a SemanticModel that ignores accessibility rules when answering semantic questions.
+        /// </summary>
+        public abstract SemanticModel WithSuppressAccessChecks();
```
I'd much rather expose this as a new optional parameter added to the existing Compilation.GetSemanticModel overloads which takes a Boolean flag for whether accessibility checks should be suppressed for the purpose of lookup which defaults to false. Perhaps bool ignoreAccessibility or suppressAccessChecks.

AnthonyDGreen commented on src/Compilers/Core/Portable/Compilation/SemanticModel.cs:
```diff
@@ -63,6 +63,16 @@ public SyntaxTree SyntaxTree
         /// <summary>
+        /// Returns true if this is a SemanticModel that ignores accessibility rules when answering semantic questions.
+        /// </summary>
+        public bool IsSuppressingAccessChecks { get; internal set; }
```
HasAccessChecksSuppressed

OmerRaviv commented:
Thanks a lot for the feedback, @AnthonyDGreen. That's indeed a much cleaner solution. Fixed.

gafter commented:
Note that this fixes #26.

gafter commented:
Note that this fixes #378

AnthonyDGreen commented:
@tmat, @KevinH-MS, @cston, @ManishJayaswal, does this aid or address any debugging/REPL scenarios that also require bypassing access checks in the SemanticModel?

mattwar commented:
A debugging scenario is freer with accessibility than the REPL scenarios we've discussed. One REPL scenario discussed was to have an option that allowed the REPL or script code to access internals as if internals-visible-to had been set. Note, a REPL might eventually be used in a debugging scenario, so this could come in to play for that.

gafter commented:
@OmerRaviv Can you please "sign" a contributor's agreement starting at https://cla.dotnetfoundation.org/

OmerRaviv commented:
@gafter I've signed the dotnetfoundation CLA on the 18th of January using the email address that's associated with my GitHub account, omerraviv@gmail.com, and got the approval the next day.

gafter commented on src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:
```diff
@@ -2747,6 +2748,13 @@ internal Conversion ClassifyConversionForCast(int position, ExpressionSyntax exp
         /// <param name="typeParameter"></param>
         public abstract ITypeParameterSymbol GetDeclaredSymbol(TypeParameterSyntax typeParameter, CancellationToken cancellationToken = default(CancellationToken));
 
+        protected BinderFlags GetSemanticModelBinderFlags()
```
This should be internal, not protected, as it should not be part of the publicly visible API.

gafter commented on src/Compilers/CSharp/Portable/Compilation/SyntaxTreeSemanticModel.cs:
```diff
@@ -29,10 +29,12 @@ internal partial class SyntaxTreeSemanticModel : CSharpSemanticModel
         {
             this.compilation = compilation;
             this.syntaxTree = syntaxTree;
+            this.HasAccessChecksSuppressed = suppressAccessChecks;
```
Please use the pattern as for other things; make this a local field in only those subclasses that need it, and make the original property a virtual (you can make the base implemenetation return false).

gafter commented on src/Compilers/VisualBasic/Portable/Binding/SemanticModelBinder.vb:
```diff
@@ -8,15 +8,27 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
-        Public Shared Function Mark(binder As Binder) As Binder
+        Public Shared Function Mark(binder As Binder, Optional ignoreAccessibility As Boolean = False) As Binder
             Return If(
                 binder.IsSemanticModelBinder,
```
This condition is insufficient to ensure the existing binder is right. You also need to test its accessibility ignorance.

gafter commented:
This pull request failed its automatic merge and test. Please re-merge with the latest and resubmit the pull request.

OmerRaviv commented:
@gafter I've addressed your latest comments and resubmitted the Pull Request as one squashed commit. I've updated the commit message to reflect the new design decisions.

gafter commented:
@tmat Tomas, can you please review this? Do you think this API change makes sense?

gafter commented:
:+1: This looks good to me.

@jaredpar @mattwar @AlekseyTs @VSadov Can you please check this too before it is pulled?

cston commented on src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:
```diff
@@ -2805,7 +2813,7 @@ public ILocalSymbol GetDeclaredSymbol(CatchDeclarationSyntax catchDeclaration, C
                 }
 
-                catchBinder = catchBinder.WithAdditionalFlags(BinderFlags.SemanticModel);
+                catchBinder = enclosingBinder.GetBinder(catchClause).WithAdditionalFlags(GetSemanticModelBinderFlags());
```
catchBinder = catchBinder.WithAdditionalFlags(GetSemanticModelBinderFlags());

tmat commented:
Looks reasonable.

@rchande @amcasey Could we use this in our EE completion provider? The EE allows evaluating internal members but completion doesn't show them, which is a bummer since the IDE actively fights you when you try to write them.

jaredpar commented on src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:
```diff
@@ -51,6 +51,7 @@ internal abstract class CSharpSemanticModel : SemanticModel
         internal abstract CSharpSyntaxNode Root { get; }
 
+
         // Is this node one that could be successfully interrogated by GetSymbolInfo/GetTypeInfo/GetMemberGroup/GetConstantValue?
```
Nit: double blank line.

cston commented on src/Compilers/VisualBasic/Portable/Binding/AttributeSemanticModel.vb:
```diff
@@ -11,15 +11,15 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
-        Private Sub New(root As VisualBasicSyntaxNode, binder As Binder, Optional parentSemanticModelOpt As SyntaxTreeSemanticModel = Nothing, Optional speculatedPosition As Integer = 0)
-            MyBase.New(root, binder, parentSemanticModelOpt, speculatedPosition)
+        Private Sub New(root As VisualBasicSyntaxNode, binder As Binder, Optional parentSemanticModelOpt As SyntaxTreeSemanticModel = Nothing, Optional speculatedPosition As Integer = 0, Optional isSupressingAccessChecks As Boolean = False)
+            MyBase.New(root, binder, parentSemanticModelOpt, speculatedPosition, isSupressingAccessChecks)
```
Consider 'suppressAccessChecks' instead for consistency with C#.

rchande commented:
@tmat Since it's just a public API on Compilation, we could have completion use this during debug time. Does this automatically work with Speculative SemanticModels?

jaredpar commented on src/Compilers/Core/Portable/Compilation/SemanticModel.cs:
```diff
@@ -63,6 +63,14 @@ public SyntaxTree SyntaxTree
         /// <summary>
+        /// Returns true if this is a SemanticModel that ignores accessibility rules when answering semantic questions.
+        /// </summary>
+        public virtual bool HasAccessChecksSuppressed
```
Can we name this `SuppressAccessChecks`?  

The pattern of `SuppressY` is used a number of times in our code base both internally and publically.  The pattern of `HasYSuppressed` does not appear anywhere I could find though we do have a few occurrences of `IsYSuppressed` but they are all methods and internal.  

That pattern seems to be more consistent with other public, and even internal, APIs that we expose.  


@gafter what do you think?

gafter replied:
What about IgnoresAccessibility?

jaredpar commented:
:+1: 

I had one bit of feedback about the choice of API name here.  I think `SuppressAccessChecks` is more consistent with existing patterns `HasAccessChecksSuppressed`.  Wanted to get others feedback on that before merging.  

Other than that though it looks great!

gafter commented:
SuppressAccessChecks  sounds a bit more like a command than a test.

@AnthonyDGreen  what do you think?

jaredpar commented:
Merging a couple of comment threads here.  The two alternate proposals for name are:

- SuppressAccessChecks
- IgnoresAccessibility

I am fine with either.  Both have precedent within the Roslyn code base and fit into existing patterns which was my primary concern. 

Small nit is I would use `IgnoreAccessibility` over `IgnoresAccessibility`.  There are no cases I could find of a plural form of ignores but plenty where the singular variety is used.

OmerRaviv commented:
@rchande Yes, this automatically supports Speculative SemanticModels, see unit tests for proof :) 
@rchande @tmat @amcasey Yes, I think it would make sense to combine this functionality with #423, so that you can get IntelliSense to include inaccessible members while debugging. 

@jaredpar @gafter @AnthonyDGreen  I don't really have an opinion regarding naming either way, please let me know once you've reached consensus.

AnthonyDGreen commented:
@jaredpar, I don't think we have that many Boolean flags in the API but in the past I've used "the if test". How does checking the flag look in an if:

```C#
if (semanticModel.HasAccessChecksSuppressed) { .. }
if (semanticModel.IgnoresAccessibility) { ... }
if (semanticModel.SuppressAccessChecks) { ... }
```
I think either of the first two read better in that context. This is also what leads me to prefer 'ignores' over 'ignore'. We had a similar discussiong when we named ControlFlowAnalysis/DataFlowAnalysis.Succeeded vs Success. If we're applying this inconsistently we should review the various cases holistically.

@terrajobst for weigh in on naming.

jaredpar commented:
@AnthonyDGreen the choice of `Ignores` vs. `Ignore` doesn't really matter to me.  All I care about is consistency in the choices.  Today we do have APIs in the singular so if we make this plural I prefer we make all of the other plural as well (even if the APIs are internal).  The same is true for `HasYSuppressed`, it breaks an existing established pattern.

gafter commented:
I can't find examples where we use the "singular" except in the imperative sense (of instructing the API to ignore something), such as LookupOptions.IgnoreAccessibility. This property test the setting.

jaredpar commented:
Here are the cases I could find.  

Singular:

- AbstractRegionDataFlowPass.IgnoreOutSemantics
- DataFlowPass.IgnoreOutSemantics
- UnassignedVariablesWarker.IgnoreOutSemantics
- IgnoreBaseClassesInLookup 
- CandidateAnalysisResult.IgnoreExtensionMethods
- ComparisonOptions.IgnoreCase
- ComparisonOptions.IgnoreAssemblyKey
- SuppressOptions.IgnoreElastic

Plural:

- Binder.IgnoresAccessibility

gafter commented:
The only ones of those that are public are options (commands) to ignore something.

jaredpar commented:
Internal or not I think we should be consistent.

JaredPar from a phone
http://blog.paranoidcoding.com/
________________________________
From: Neal Gafter <notifications@github.com>
Sent: Wednesday, February 18, 2015 2:26:37 PM
To: dotnet/roslyn
Cc: Jared Parsons
Subject: Re: [roslyn] Added “SemanticModel.WithSuppressAccessChecks” extension method. (#26)


The only ones of those that are public are options (commands) to ignore something.

—
Reply to this email directly or view it on GitHub<https://github.com/dotnet/roslyn/pull/26#issuecomment-74961945>.

AnthonyDGreen commented:
They're in different contexts. One is a flag on SemanticModel describing a _property_ of the semantic model. The Ignores isn't plural, it's third person singular. I ignore, you ignore, he/she/it _ignores_. This is the same pattern we use for SemanticModel.IsSpeculativeSemanticModel method (or Object.Equals). I am, you are, he/she/it _is_.

This is a different grammatical context than the parameter which one passes in to GetSemanticModel to cause accessibility to be ignored - ignoreAccessibility. That's describing an _action_ that should be taken (sort of a second person command) and options follow that same convention.

The problem is that depending on where the code is written a piece of data will need to be specified or described in different persons. To the person acquiring the SemanticModel it's 2nd - ignoreAccessibility, within the SemanticModel it would be 1st - IgnoreAccessibility, and outside of the SemanticModel it will be in the 3rd - IgnoresAccessibility. Things get even muddier when properties are read-write because getting is usually in the 3rd and setting is usually in the 2nd. We can't even just standardize on one because - Object.Equals is 3rd person, and IsSpeculativeSemanticModel is 3rd person. I'm not even sure what the latter would be in the 1st or 2nd persons - BeSpeculative, Speculate, AmSpeculative, or just Speculative?

So which name is used where depends heavily on who the primary consumer is. The parameters and options are primarily commands for someone else (e.g. the Compilation, the parser) and should be written in the 2nd person, and the properties on SemanticModel are primarily attributes that should be written in the 3rd person.

I agree that the internal cases should be made consistent if they are in a context where that reasoning applies.

gafter commented:
I am going to accept this pull request as-is and then submit a separate change to rename the public member. Please say something now if that sounds like a bad plan.

jaredpar commented:
Thanks fine by me.

terrajobst commented:
Sorry to be super later but hey -- better late than never :-)

> @terrajobst for weigh in on naming.

I'm with Jared. Consistency is key. As far as imperative vs. descriptive style goes: we have both. It usually depends on context. In general, it's good to design booleans to make them read nice in `if` checks. I believe the ones discussed all satisfy this:

```csharp
if (semanticModel.HasAccessChecksSuppressed) { .. }
if (semanticModel.IgnoresAccessibility) { ... }
if (semanticModel.SuppressAccessChecks) { ... }
```
